src/Bicep.Core/TypeSystem/Providers/ResourceTypeProviderFactory.cs

```diff
@@ -42,35 +44,26 @@ public ResultWithDiagnostic<IResourceTypeProvider> GetResourceTypeProviderFromFi
                 return new(cachedResourceTypeLoaders[key]);
             }
 
-            // is never null since provider restore success is validated prior.
-            var typesTgzPath = Uri.UnescapeDataString(providerDescriptor.TypesBaseUri?.AbsolutePath ?? throw new UnreachableException("the provider directory doesn't exist"));
-            var typesParentPath = Path.GetDirectoryName(typesTgzPath) ?? throw new UnreachableException("the provider directory doesn't exist");
-
-            // compose the path to the OCI manifest based on the cache root directory and provider version
-            var ociManifestPath = Path.Combine(typesParentPath, "manifest");
-            if (!fileSystem.File.Exists(ociManifestPath))
+            IResourceTypeProvider newResourceTypeLoader;
+            try
             {
-                // always exists since provider restore was successful
-                throw new UnreachableException("the provider manifest path doesn't exist");
+                if (providerDescriptor.Name != AzNamespaceType.BuiltInName)
+                {
+                    // Note (asilverman): the line of code below is meant for 3rd party provider resolution logic which is not yet implemented.
+                    throw new NotImplementedException($"Provider {providerDescriptor.Name} not supported.");
+                }
+                newResourceTypeLoader =
+                new AzResourceTypeProvider(
+                    new AzResourceTypeLoader(
+                        OciTypeLoader.FromDisk(fileSystem, providerDescriptor.TypesBaseUri)),
+                        providerDescriptor.Version);
             }
-
-            // Read the OCI manifest
-            var manifestFileContents = fileSystem.File.ReadAllText(ociManifestPath);
-            OciManifest? ociManifest = JsonConvert.DeserializeObject<OciManifest>(manifestFileContents);
-
-            if (ociManifest is null)
+            catch (Exception ex)
             {
-                return new(x => x.ErrorOccurredReadingFile(ociManifestPath)); ;
+                var invalidArtifactException = ex.As<InvalidArtifactException?>() ?? new InvalidArtifactException(ex.Message, ex, InvalidArtifactExceptionKind.NotSpecified);
```

Comment:
It looks like the only reason to create the `InvalidArtifactException` here is for string formatting. This is a bit roundabout. I would instead add a static method to `InvalidArtifactException`:

```csharp
string GetInvalidArtifactMessage(string innerMessage)
    => $"The OCI artifact is not a valid Bicep artifact. {innerMessage}";
```

Reply:
There is one more nuanced, InvalidArtifactException has a member Kind which is set to InvalidArtifactExceptionKind.NotSpecified to indicate that something unexpected about the artifact creation took place, there are many operations that may throw and this is a last effort attempt to catch them to emit a diagnostic.

Reply:
You're not using it here though, you're just using the exception to instantly return the message (which doesn't depend on kind).